lib/core/include/irods/rodsErrorTable.h

```diff
@@ -243,6 +243,8 @@ NEW_ERROR(RESOLUTION_ERROR,                            -174000)
 NEW_ERROR(INVALID_HANDLE,                              -175000)
 NEW_ERROR(DOES_NOT_EXIST,                              -176000)
 NEW_ERROR(TYPE_NOT_SUPPORTED,                          -177000)
+NEW_ERROR(ILLEGAL_ACCESS_VIOLATION,                    -178000)
+NEW_ERROR(AUTHENTICATION_ERROR,                        -179000)
```

Comment:
Notice the new error codes.

Reply:
@trel @alanking Any thoughts on these names?

Reply:
I realize the direction we are headed is more "generic" error codes, so, I don't really have anything better to suggest for these. They seem fine.

`ILLEGAL_ACCESS_VIOLATION` seemed to be used mainly when attempting to dereference a nullptr for the type-casting functions, right? I was tempted to suggest something like `NULLPTR_DEREFERENCE` but it doesn't communicate any more information than what you have - that is, it is a subset of the existing name you have now.

`AUTHENTICATION_ERROR` is okay as long as `clientLogin` or whatever is returning a little more information in the `rError` stack or something (but... maybe not *too much* information...)

Reply:
What info would you like to see returned in the rError stack on a failed authentication attempt?

Perhaps the following ...
- username#zone
- hostname
- zone

Or is there other info you feel should be included?

Reply:
don't we already have an equivalent to `AUTHENTICATION_ERROR`?   or maybe we would transition the older one to this one over time...

Reply:
> `ILLEGAL_ACCESS_VIOLATION` is a weird ... phrase... is one of these better? `ILLEGAL_ACCESS` or `ILLEGAL_MEMORY_ACCESS` or `ACCESS_VIOLATION`

`ILLEGAL_ACCESS` or `ACCESS_VIOLATION` work.

I think I like `ILLEGAL_ACCESS`, but if we already use `VIOLATION` somewhere, then I think that wins.

I'll verify this and make adjustments.

---

As for the auth error code, we have `CAT_INVALID_AUTHENTICATION`, but that prefix bothers me because it is catalog-specific and that may not be true for every case. This new code is generic enough that it can be used across all auth error cases.

Reply:
these seem most useful.
- username#zone
- connected server hostname
- perhaps also server hostname of provider
- type of auth attempted?

Reply:
I'm not saying that `AUTHENTICATION_ERROR` has to do any more work, but `clientLogin` may have messages and/or error codes from the auth plugins that may be masked by throwing this. Just want to make sure that the intended communications make it back to the user. Having more information probably isn't a bad thing

Reply:
Okay. I'll capture this conversation in a new issue so this PR can be merged.

Making those changes will apply to both connection classes and should be trivial to implement.

Reply:
We don't use `ILLEGAL` or `VIOLATION` in any error code names.

Either would work, but I think `ILLEGAL` works better in that it would be used as a prefix on the error code name. That would allow similar errors to be grouped. For example ...
- `ILLEGAL_ACCESS`
- `ILLEGAL_OPERATION`
- `ILLEGAL_MODIFICATION`

Do either of you have a preference on this?

Reply:
`ILLEGAL_ACCESS_VIOLATION` seems better to me. "Violation" is pretty commonly used for these types of situations in C/C++ land, and leaving it off could be conflated with "access" to, say, a data object, or something else in the iRODS system. Just my two cents

Reply:
Good point. I stole those words from Java. I think they have an exception that uses those words. Or maybe that was nfs4j.

Reply:
I think I also see @trel's point about that name.

@trel Correct me if I'm wrong, but are you getting at the use of **illegal** and **violation** being used in the same name?

I guess he's saying they mean the same thing so why have both.

Reply:
Alternatively, how do we feel about `ILLEGAL_ACCESS_ERROR`?

That is equivalent to java's `IllegalAccessError` exception. See https://docs.oracle.com/javase/8/docs/api/java/lang/IllegalAccessError.html.

Reply:
Well, I'm not a big fan of `_ERROR` since these are error codes.

Reply:
yeah, no `_ERROR`

Reply:
> We don't use `ILLEGAL` or `VIOLATION` in any error code names.
> 
> Either would work, but I think `ILLEGAL` works better in that it would be used as a prefix on the error code name. That would allow similar errors to be grouped. For example ...
> 
> * `ILLEGAL_ACCESS`
> * `ILLEGAL_OPERATION`
> * `ILLEGAL_MODIFICATION`
> 
> Do either of you have a preference on this?

illegal in this sense... means... invalid?  or not allowed due to permissions?

Reply:
It is due to the rules of the class. You get this error if you attempt to call a function and the pre-condition for calling it is violated.

For example ...
```cpp
client_connection conn;                      // connect to irods.
auto other_conn = std::move(conn);           // transfer ownership of the RcComm to other_conn.
auto& conn_ref = static_cast<RcComm&>(conn); // BOOM! conn doesn't have a valid RcComm.
```

Reply:
After some discussion, @alanking and I determined that we should drop the `ILLEGAL_ACCESS_VIOLATION` error code and use `SYS_LIBRARY_ERROR`.

The reason for this is that iRODS error codes are defined for users of the iRODS system, not developers. And, it could be misleading to return such error codes.

This also means we need to define rules around what constitutes a new error code.

Reply:
oh, well that's nice.

Reply:
Done.